README.md
# Amazon "App Runner Service Deploy" action for GitHub Actions

Registers an AWS App Runner Service and deploys the application using the source code of a given GitHub repository. Supports both source code and Docker image based service.

## Table of Contents

<!-- toc -->

- [V2 changes](#v2-changes)
- [Usage](#usage)
  - [Code based service](#code-based-service)
  - [Image based service](#image-based-service)
- [Credentials and Region](#credentials-and-region)
- [Permissions](#permissions)
- [Troubleshooting](#troubleshooting)
- [License Summary](#license-summary)
- [Security Disclosures](#security-disclosures)

<!-- tocstop -->

## V2 Changes

This action's codebase has been refactored to support future growth as well as to simplify to processes of adding new capabilities.

Refer to the [Changelog](./CHANGELOG.md) for the change history.

## Usage

This github action supports two types of App Runner services: source code based and docker image based.

See [action.yml](action.yml) for the full documentation for this action's inputs and outputs.

### Code based service

Source code is application code that App Runner builds and deploys for you. You point App Runner to a source code repository and choose a suitable runtime. App Runner builds an image that's based on the base image of the runtime and your application code. It then starts a service that runs a container based on this image.

> Note: The list of supported runtime platforms is available [here](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-apprunner/enums/runtime.html).

Here is the sample for deploying a NodeJS based service:

```yaml
name: Deploy to App Runner 
on:
  push:
    branches: [main] # Trigger workflow on git push to main branch
  workflow_dispatch: # Allow manual invocation of the workflow

jobs:  
  deploy:
    runs-on: ubuntu-latest
    # These permissions are needed to interact with GitHub's OIDC Token endpoint.
    permissions:
      id-token: write
      contents: read
    
    steps:            
      - name: Configure AWS credentials
        uses: aws-actions/configure-aws-credentials@v1-node16
        with:
          # Use GitHub OIDC provider
          role-to-assume: ${{ secrets.AWS_ASSUME_ROLE_ARN }}
          aws-region: ${{ secrets.AWS_REGION }}
          
      - name: Deploy to App Runner
        id: deploy-apprunner
        uses: awslabs/amazon-app-runner-deploy@main
        env:
          SERVER_PORT: 80
          SECRET_ENV: secret_env
        with:
          service: app-runner-git-deploy-service
          source-connection-arn: ${{ secrets.AWS_CONNECTION_SOURCE_ARN }}
          repo: https://github.com/${{ github.repository }}
          branch: ${{ github.ref }}
          runtime: NODEJS_16
          build-command: npm install
          start-command: npm start
          port: 18000
          region: ${{ secrets.AWS_REGION }}
          cpu : 1
          memory : 2
          # Deprecated: wait-for-service-stability: true
          # The new way to control service stability timeout
          wait-for-service-stability-seconds: 600
          copy-env-vars: |
            SERVER_PORT
          copy-secret-env-vars: |
            SECRET_ENV
          instance-role-arn: ${{ secrets.INSTANCE_ROLE_ARN }}
          tags: >
            { "env": "test" }
      
      - name: App Runner URL
        run: echo "App runner URL ${{ steps.deploy-apprunner.outputs.service-url }}" 
```

**Note:**

- **AWS_CONNECTION_SOURCE_ARN** is the ARN of the source code connector in AWS App Runner, for more details refer to this [documentation](https://docs.aws.amazon.com/apprunner/latest/dg/manage-connections.html)

### Image based service

Here, a source image (*that could be a public or private container image stored in an image repository*) can get used by App Runner to get the service running on a container. No build stage is necessary. Rather, you provide a ready-to-deploy image.

Here is the sample for deploying a App Runner service based on docker image:

```yaml
name: Deploy to App Runner
on:
  push:
    branches: [main] # Trigger workflow on git push to main branch
  workflow_dispatch: # Allow manual invocation of the workflow

jobs:  
  deploy:
    runs-on: ubuntu-latest
    # These permissions are needed to interact with GitHub's OIDC Token endpoint.
    permissions:
      id-token: write
      contents: read
    
    steps:      
      - name: Checkout
        uses: actions/checkout@v2
        with:
          persist-credentials: false
          
      - name: Configure AWS credentials
        id: aws-credentials
        uses: aws-actions/configure-aws-credentials@v1-node16
        with:
          # Use GitHub OIDC provider
          role-to-assume: ${{ secrets.AWS_ASSUME_ROLE_ARN }}
          aws-region: ${{ secrets.AWS_REGION }}

      - name: Login to Amazon ECR
        id: login-ecr
        uses: aws-actions/amazon-ecr-login@v1        

      - name: Build, tag, and push image to Amazon ECR
        id: build-image
        env:
          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
          ECR_REPOSITORY: nodejs
          IMAGE_TAG: ${{ github.sha }}
        run: |
          docker build -t $ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG .
          docker push $ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG
          echo "::set-output name=image::$ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG"  
          
      - name: Deploy to App Runner Image
        id: deploy-apprunner
        uses: awslabs/amazon-app-runner-deploy@main
        with:
          service: app-runner-git-deploy-service
          image: ${{ steps.build-image.outputs.image }}
          access-role-arn: ${{ secrets.ROLE_ARN }}
          region: ${{ secrets.AWS_REGION }}
          cpu : 1
          memory : 2
          # Deprecated: wait-for-service-stability: true
          # The new way to control service stability timeout
          wait-for-service-stability-seconds: 1200
      
      - name: App Runner URL
        run: echo "App runner URL ${{ steps.deploy-apprunner.outputs.service-url }}" 
```

**Note:**

- The above example uses github action, to build the docker image, push it to AWS ECR and use the same for App Runner deployment
- **ROLE_ARN** is the Amazon Resource Name (ARN) of the IAM role that grants the App Runner service access to a source repository. It's required for ECR image repositories (but not for ECR Public repositories)

## Credentials and Region

This action relies on the [default behavior of the AWS SDK for Javascript](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html) to determine AWS credentials and region.
Use [the `aws-actions/configure-aws-credentials` action](https://github.com/aws-actions/configure-aws-credentials) to configure the GitHub Actions environment with environment variables containing AWS credentials and your desired region.

We recommend using [GitHub's OIDC provider](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services) to get short-lived credentials needed for your actions. 

It is recommended to follow [Amazon IAM best practices](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html) for the AWS credentials used in GitHub Actions workflows, including:

- Do not store credentials in your repository's code.  You may use [GitHub Actions secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets) to store credentials and redact credentials from GitHub Actions workflow logs.
- [Create an individual IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#create-iam-users) with an access key for use in GitHub Actions workflows, preferably one per repository. Do not use the AWS account root user access key.
- [Grant least privilege](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#grant-least-privilege) to the credentials used in GitHub Actions workflows.  Grant only the permissions required to perform the actions in your GitHub Actions workflows.  See the Permissions section below for the permissions required by this action.
- [Rotate the credentials](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#rotate-credentials) used in GitHub Actions workflows regularly.
- [Monitor the activity](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#keep-a-log) of the credentials used in GitHub Actions workflows.

## Permissions

Generally this action requires the following minimum set of permissions:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "apprunner:ListServices",
                "apprunner:CreateService",
                "apprunner:UpdateService",
                "apprunner:DescribeService",
                "apprunner:TagResource",
                "iam:PassRole"
            ],
            "Resource": "*"
        }
    ]
}
```

For **Image based service** this action requires additionally the following minimum set of permissions:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "ecr:DescribeImages",
                "ecr:GetAuthorizationToken",
                "ecr:BatchCheckLayerAvailability"
            ],
            "Resource": "*"
        }
    ]
}
```

## Troubleshooting

This action emits debug logs to help troubleshoot deployment failures.  To see the debug logs, create a secret named `ACTIONS_STEP_DEBUG` with value `true` in your repository.

## License Summary

This code is made available under the MIT-0 license, for details refer to [LICENSE](LICENSE) file.

## Security Disclosures

If you would like to report a potential security issue in this project, please do not create a GitHub issue.  Instead, please follow the instructions [here](https://aws.amazon.com/security/vulnerability-reporting/) or [email AWS security directly](mailto:aws-security@amazon.com).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.5.2] - 2023-10-30

### Changed

- NPM dependency updates

## [2.5.1] - 2023-10-10

### Changed

- NPM dependency updates

## [2.5.0] - 2023-06-14

### Changed

- Allow using float value in memory and cpu config ([#42](https://github.com/awslabs/amazon-app-runner-deploy/pull/42))
- NPM dependency updates

## [2.4.0] - 2023-04-27

### Added

- Support RuntimeEnvironmentSecrets ([#29](https://github.com/awslabs/amazon-app-runner-deploy/issues/29))
- Support InstanceRoleArn ([#40](https://github.com/awslabs/amazon-app-runner-deploy/issues/40))

### Changed

- NPM dependency updates

## [2.3.0] - 2023-04-19

### Added

- Action will fail if the deployment was rolled back due to an error ([#11](https://github.com/awslabs/amazon-app-runner-deploy/issues/11), [#32](https://github.com/awslabs/amazon-app-runner-deploy/issues/32))

## [2.2.2] - 2023-03-30

### Added

- Action version logging

### Changed

- Update all external NPM dependencies

## [2.2.1] - 2023-03-29

### Fixed

- Error: The auto scaling configuration ARN you specified isn't valid. ARN: '' ([#36](https://github.com/awslabs/amazon-app-runner-deploy/issues/36))

## [2.2.0] - 2023-03-28

### Added

- Optional `auto-scaling-config-arn` input. When specified this will be applied to new and existing services.

## [2.1.0] - 2023-02-08

### Added

- Tagging support ([#21](https://github.com/awslabs/amazon-app-runner-deploy/issues/21))

## [2.0.1] - 2023-01-10

- Dependency updates

## [2.0.0] - 2022-10-02

### Added

- Action outputs: `service-arn` and `service-url`.
- Optional `wait-for-service-stability-seconds` configuration parameter. Valid range is between 10 and 3600 seconds. The default is 600 seconds if `wait-for-service-stability` flag is set to true. This is a replacement for the existing `wait-for-service-stability` boolean flag, which will be deprecated in a future release.
- Optional `action` parameter. The only valid value is `create_or_update` (case insensitive) and the parameter is optional. It is introduced for adding more sub-actions in a future release.
- Support for additional `runtime` parameter values: `DOTNET_6`, `GO_1`, `NODEJS_16`, `PHP_81`, `RUBY_31`.
- If there is an existing service with CREATE_FAILED status, it will be deleted first, before the new service creation is attempted.
- [Add support for environment variables](https://github.com/awslabs/amazon-app-runner-deploy/issues/4).

### Changed

- **BREAKING CHANGE**: The default branch name is now `main`.
- **BREAKING CHANGE**: Action migrated to run on Node16.
- **BREAKING CHANGE**: Update all NPM module references, including major version upgrades.
- **BREAKING CHANGE**: Refactor code to be compatible with future enhancements and updates.
- **BREAKING CHANGE**: Log messages changed to match the new refactored code structure.
- Supported runtime list is no longer hardcoded, but automatically synchronized with the one, defined by [AppRunner Client SDK](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-apprunner/enums/runtime.html), so that new runtime identifier will become available with AWS SDK NPM module updates.
- Mark `wait-for-service-stability` as obsolete.

### Fixed

- [Image is ignored for existing App Runner Service](https://github.com/awslabs/amazon-app-runner-deploy/issues/13)
- [Runtime Support for Node.js 16](https://github.com/awslabs/amazon-app-runner-deploy/issues/10)

## [1.x]

The initial family of releases with core functionality.


src/action-configuration.ts
import { getInput, getMultilineInput, InputOptions } from "@actions/core";
import { Runtime, Tag } from "@aws-sdk/client-apprunner";

// supported GitHub action modes
enum Actions {
    // Create a new service or update an existing one
    CreateOrUpdate = 'create_or_update',
}

export interface ICodeConfiguration {
    sourceType: 'code';
    repoUrl: string;
    branch: string;
    sourceConnectionArn: string;
    runtime: Runtime;
    buildCommand: string;
    startCommand: string;
}

export interface IImageConfiguration {
    sourceType: 'image';
    imageUri: string;
    accessRoleArn: string;
}

export interface ICreateOrUpdateActionParams {
    action: Actions.CreateOrUpdate;
    serviceName: string;
    sourceConfig: ICodeConfiguration | IImageConfiguration;
    port: number;
    waitForService: boolean;
    waitTimeout: number;
    region: string;
    cpu: number;
    memory: number;
    environment?: Record<string, string>;
    environmentSecret?: Record<string, string>;
    tags: Tag[]
    autoScalingConfigArn?: string;
    instanceRoleArn?: string;
}

export type IActionParams = ICreateOrUpdateActionParams;

interface IValidationRules {
    min?: number;
    max?: number;
}

function getOptionalInputNumber(name: string, { validation, intOnly }: { validation?: IValidationRules, intOnly?: boolean } = {}): number | undefined {
    const val = getInput(name, { required: false, trimWhitespace: true });
    if (!val) {
        return undefined;
    }

    const result = intOnly
      ? Number.parseInt(val)
      : Number.parseFloat(val);

    if (isNaN(result)) {
        throw new Error(`${name} value is not a valid number: ${val}`);
    }

    if (validation?.min && validation.min > result) {
        throw new Error(`${name} value (${result}) is less then the allowed minimum (${validation.min})`);
    }

    if (validation?.max && validation.max < result) {
        throw new Error(`${name} value (${result}) is greater then the allowed maximum (${validation.max})`);
    }

    return result;
}

function getInputNumber(name: string, defaultValue: number, { validation, intOnly }: { validation?: IValidationRules, intOnly?: boolean } = {}): number {
    return getOptionalInputNumber(name, { validation, intOnly }) ?? defaultValue;
}

function getInputStr(name: string, defaultValue: string): string {
    return getInput(name, { required: false, trimWhitespace: true }) || defaultValue;
}

function getOptionalInputStr(name: string, options?: InputOptions): string | undefined {
    const value = getInput(name, { required: false, ...options })
    return (value.length > 0) ? value : undefined
}

function getInputBool(name: string, defaultValue: boolean): boolean {
    const val = getInput(name, { required: false, trimWhitespace: true });
    if (!val) {
        return defaultValue;
    }

    return ['1', 'true'].includes(val.toLowerCase());
}

export function getConfig(): IActionParams {
    const rawActionInput = getInput('action', { required: false, trimWhitespace: true });

    switch (rawActionInput.toLowerCase() || Actions.CreateOrUpdate) {
        case Actions.CreateOrUpdate:
            return getCreateOrUpdateConfig();
        default:
            throw new Error(`Unsupported action: ${rawActionInput}`);
    }
}

function getCreateOrUpdateConfig(): ICreateOrUpdateActionParams {
    const action = Actions.CreateOrUpdate;
    // Service name - required input with no default value
    const serviceName = getInput('service', { required: true, trimWhitespace: true });

    // Port number - 80
    const port = getInputNumber('port', 80, { intOnly: true });

    // Region - us-east-1
    const region = getInputStr('region', 'us-east-1');

    // Wait for service to complete the creation/update - false
    const waitForService = getInputBool('wait-for-service-stability', false);
    const waitTimeout = getOptionalInputNumber('wait-for-service-stability-seconds', { validation: { min: 10, max: 3600 }, intOnly: true });

    // CPU - 1 vCPU
    const cpu = getInputNumber('cpu', 1);

    // Memory - 2GB
    const memory = getInputNumber('memory', 2);

    // Source docker image URL - this will switch between deploying source code or docker image
    const imageUri = getInput('image', { required: false, trimWhitespace: true });

    const envVarNames = getMultilineInput('copy-env-vars', { required: false });

    const secretEnvVarNames = getMultilineInput('copy-secret-env-vars', { required: false });

    const tags = getInput('tags', { required: false })

    const autoScalingConfigArn = getOptionalInputStr('auto-scaling-config-arn', { trimWhitespace: true });

    const instanceRoleArn = getOptionalInputStr('instance-role-arn', { trimWhitespace: true });

    return {
        action,
        serviceName,
        region,
        port,
        waitForService: waitForService || !!waitTimeout,
        waitTimeout: waitTimeout ?? 600,
        cpu,
        memory,
        sourceConfig: imageUri ? getImageConfig(imageUri) : getSourceCodeConfig(),
        environment: getEnvironmentVariables(envVarNames),
        environmentSecret: getEnvironmentVariables(secretEnvVarNames),
        tags: getTags(tags),
        autoScalingConfigArn: autoScalingConfigArn,
        instanceRoleArn: instanceRoleArn,
    };
}

function getBranch(): string {
    // Breaking change - default branch name switched to `main`!!!
    const branch = getInput('branch', { required: false }) || 'main';

    return (branch.startsWith("refs/")) ? branch.split("/")[2] : branch;
}

function getImageConfig(imageUri: string): IImageConfiguration {
    if (getInput('repo', { required: false })) {
        throw new Error('Either docker image registry or code repository expected, not both');
    }
    return {
        sourceType: 'image',
        imageUri,
        accessRoleArn: getInput('access-role-arn', { required: true }),
    };
}

function getSourceCodeConfig(): ICodeConfiguration {
    return {
        sourceType: 'code',
        sourceConnectionArn: getInput('source-connection-arn', { required: true }),
        repoUrl: getInput('repo', { required: true }),
        branch: getBranch(),
        runtime: getRuntime(),
        buildCommand: getInput('build-command', { required: true }),
        startCommand: getInput('start-command', { required: true }),
    }
}

function getRuntime(): Runtime {
    const rawRuntime = getInput('runtime', { required: true });
    const runtime = rawRuntime.toUpperCase();
    if (!Object.keys(Runtime).includes(runtime)) {
        throw new Error(`Specified runtime (${rawRuntime}) does not belong to the supported range: ${JSON.stringify(Object.keys(Runtime))}`);
    }

    return Runtime[<keyof typeof Runtime>runtime];
}

function getEnvironmentVariables(envVarNames: string[]): Record<string, string> | undefined {
    if (envVarNames.length > 0) {
        const mapped = envVarNames.reduce((acc: Record<string, string>, env) => {
            const envVarValue = process.env[env];
            if (envVarValue !== undefined) {
                acc[env] = envVarValue;
            }
            return acc;
        }, {});
        if (Object.keys(mapped).length > 0) {
            return mapped;
        }
    }
}

function getTags(tags: string): Tag[] {
  if (!tags.length) {
    return []
  }

  const parsed = JSON.parse(tags);
  return Object.keys(parsed).reduce((acc, tagKey) => {
    return [
      ...acc,
      {
        Key: tagKey,
        Value: parsed[tagKey]
      }
    ]
  }, [] as Tag[])
}
